hazelcast/src/main/java/com/hazelcast/instance/AbstractAddressPicker.java

```diff
@@ -0,0 +1,141 @@
+
+package com.hazelcast.instance;
+
+import com.hazelcast.config.NetworkConfig;
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.IOUtil;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.nio.channels.ServerSocketChannel;
+import java.util.concurrent.TimeUnit;
+
+abstract class AbstractAddressPicker implements AddressPicker {
+    private static final int SOCKET_TIMEOUT_MILLIS = (int) TimeUnit.SECONDS.toMillis(1);
+    private static final int SOCKET_BACKLOG_LENGTH = 100;
+
+    final ILogger logger;
+    private final NetworkConfig networkConfig;
+
+    private ServerSocketChannel serverSocketChannel;
+
+    AbstractAddressPicker(NetworkConfig networkConfig, ILogger logger) {
+        this.networkConfig = networkConfig;
+        this.logger = logger;
+    }
+
+    /**
+     * Creates and binds {@code ServerSocketChannel} using {@code bindAddress} and {@code initialPort}.
+     * <p>
+     * If the {@code initialPort} is in use, then an available port can be selected by doing an incremental port
+     * search if {@link NetworkConfig#isPortAutoIncrement()} allows.
+     * <p>
+     * When both {@code initialPort} and {@link NetworkConfig#getPort()} are zero, then an ephemeral port will be used.
+     * <p>
+     * When {@code bindAny} is {@code false}, {@code ServerSocket} will be bound to specific {@code bindAddress}.
+     * Otherwise, it will be bound to any local address ({@code 0.0.0.0}).
+     *
+     * @param bindAddress InetAddress to bind created {@code ServerSocket}
+     * @param initialPort Initial port number to attempt to bind
+     * @param bindAny     Flag to decide whether to bind given {@code bindAddress} or any local address
+     * @return actual port number that created {@code ServerSocketChannel} is bound to
+     */
+    final int createServerSocketChannel(InetAddress bindAddress, int initialPort, boolean bindAny) {
+        int portCount = networkConfig.getPortCount();
+        boolean portAutoIncrement = networkConfig.isPortAutoIncrement();
+        logger.finest("inet reuseAddress:" + networkConfig.isReuseAddress());
+
+        if (initialPort == 0) {
+            initialPort = networkConfig.getPort() ;
+        }
+        if (initialPort == 0) {
+            logger.info("No explicit port is given, system will pick up an ephemeral port.");
+        }
+        int portTrialCount = initialPort > 0 && portAutoIncrement ? portCount : 1;
+
+        try {
+            return tryOpenServerSocketChannel(bindAddress, initialPort, portTrialCount, bindAny);
+        } catch (IOException e) {
+            String message = "Cannot bind to a given address: " + bindAddress + ". Hazelcast cannot start. ";
+            if (networkConfig.isPortAutoIncrement()) {
+                message += "Config-port: " + networkConfig.getPort() + ", latest-port: " + (initialPort + portTrialCount);
```

Comment:
Should this be `initialPort + portTrialCount - 1`?

Reply:
right. 👍